### PR TITLE
fix: doc_links link type - Determine link type at start (fixes #12601)

### LINKED ISSUE #12601: Rustdoc markdown comments with autolinks (links with angular brackets) are broken
**rust-analyzer version**: 0.0.0 (427061da1 2022-06-19)

**rustc version**: rustc 1.61.0 (fe5b13d68 2022-05-18)

**relevant settings**: None

When a Rust markdown comment with a link of the form `<https://example.com>` is added to a method or struct etc, on hover in vscode, the resulting link cannot be clicked and shows up with mismatched brackets as `<https://example.com](https://example.com)`.

I was able to find a likely source of the bug, but I don't currently have time to set up my development environment and work on a pull request myself, but I hope this info is useful. In the [map_links function in doc_links.rs](https://github.com/rust-lang/rust-analyzer/blob/439a513efb77b956ed318cef6a4af421054d29e8/crates/ide/src/doc_links.rs#L357), when a link type is changed in the callback, the "End" event's link type changes, but the "Start" event's link type is unaffected, causing a mismatch, which `cmark_resume_with_options` does not reconcile.

For example, in the `rewrite_links` function in `doc_links.rs`, after the call to `map_links`, the value of `doc` for the example above is as follows:
```
Event: Start(Paragraph)
Event: Start(Link(Autolink, Borrowed("https://example.com"), Borrowed("")))
Event: Text(Boxed("https://example.com"))
Event: End(Link(Inline, Boxed("https://example.com"), Borrowed("")))       
Event: End(Paragraph)
```
I believe the "Inline"/"Autolink" mismatch is the main source of this bug.

### PR DESCRIPTION
**Pull request by erhuve**
fixes #12601
Looked like autolink/inline mismatch happened because end_link_type was parsed from the Text/Code events.
I changed it to be determined from the Start event, which should hopefully remain accurate while staying true to the cases where link type may need to be changed, according to the comment 
```
// normally link's type is determined by the type of link tag in the end event,
// however in some cases we want to change the link type, for example,
// `Shortcut` type doesn't make sense for url links
```
Hopefully this is the desired behavior?
Untitled

### CODE REVIEW

**Comment by erhuve** [comment]
I reintroduced setting the end link type in Text/Code but with AutoLink as an edge case. It does feel a little rough to do that; I'll try to think more about a nicer solution and to understand the full range of cases better.

File: crates/ide/src/doc_links.rs [hunk 1]
```diff
@@ -362,14 +362,15 @@ fn map_links<'e>(
     // holds the origin link target on start event and the rewritten one on end event
     let mut end_link_target: Option<CowStr> = None;
     // normally link's type is determined by the type of link tag in the end event,
-    // however in same cases we want to change the link type, for example,
+    // however in some cases we want to change the link type, for example,
     // `Shortcut` type doesn't make sense for url links
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:366]: This comment actually confuses me, the code below makes no mention of Shortcut? I wonder if this is outdated?

Comment (Veykril) [reply crates/ide/src/doc_links.rs:366]: Ah I see, the comment is phrased a bit confusingly I guess.

File: crates/ide/src/doc_links.rs [hunk 2]
```diff
@@ -384,14 +385,18 @@ fn map_links<'e>(
             let (link_type, link_target_s, link_name) =
                 callback(&end_link_target.take().unwrap(), &s);
             end_link_target = Some(CowStr::Boxed(link_target_s.into()));
-            end_link_type = link_type;
+            if !matches!(end_link_type, Some(link) if link == LinkType::Autolink) {
```
Comment (Veykril) [inline crates/ide/src/doc_links.rs:388]:
```suggestion
            if !matches!(end_link_type, Some(LinkType::Autolink)) {
```
LinkType is be `StructuralPartialEq` so this should work I believe, same for the check below

File: crates/ide/src/doc_links.rs (hunk 1 above)
Comment (yue4u) [reply crates/ide/src/doc_links.rs:366]: Oh thanks for the fix, I missed the the autolink case in the pr before.

>  the code below makes no mention of Shortcut

I meant to say the `Shortcut` types coming from original start/end events, which are deconstructed to `link_type` in the following code. Let's make the comment more clearer?

**Comment by yue4u** [comment]
> I reintroduced setting the end link type in Text/Code but with AutoLink as an edge case. It does feel a little rough to do that; I'll try to think more about a nicer solution and to understand the full range of cases better.

Yes, I just followed the origin code before https://github.com/rust-lang/rust-analyzer/pull/12150 and tried to avoid heavy refactoring but this approach is indeed somewhat unnatural.

File: crates/ide/src/doc_links.rs (hunk 1 above)
Comment (erhuve) [reply crates/ide/src/doc_links.rs:366]: Agreed, will do these later today when I'm home

**Comment by Veykril** [comment]
@bors delegate+
With a small test added and the comment updated this looks good to me. Thanks!

**Pull request merged**